types.go

```diff
@@ -438,25 +438,54 @@ func (rr *TXT) String() string { return rr.Hdr.String() + sprintTxt(rr.Txt) }
 
 func sprintName(s string) string {
 	var dst strings.Builder
-	dst.Grow(len(s))
+
 	for i := 0; i < len(s); {
 		if i+1 < len(s) && s[i] == '\\' && s[i+1] == '.' {
-			dst.WriteString(s[i : i+2])
+			if dst.Len() != 0 {
+				dst.WriteString(s[i : i+2])
+			}
 			i += 2
 			continue
 		}
 
 		b, n := nextByte(s, i)
-		switch {
-		case n == 0:
-			i++ // dangling back slash
-		case b == '.':
-			dst.WriteByte('.')
+		if n == 0 {
+			i++
+			continue
+		}
+		if b == '.' {
+			if dst.Len() != 0 {
+				dst.WriteByte('.')
+			}
+			i += n
+			continue
+		}
+		switch b {
+		case ' ', '\'', '@', ';', '(', ')', '"', '\\': // additional chars to escape
+			if dst.Len() == 0 {
+				dst.Grow(len(s) * 2)
+				dst.WriteString(s[:i])
+			}
+			dst.WriteByte('\\')
+			dst.WriteByte(b)
 		default:
-			writeDomainNameByte(&dst, b)
+			if ' ' <= b && b <= '~' {
+				if dst.Len() != 0 {
+					dst.WriteByte(b)
+				}
+			} else {
+				if dst.Len() == 0 {
+					dst.Grow(len(s) * 2)
```

Comment:
some question here why the `*2`?

Reply:
> The `* 2` is because we need more than `len(s)` bytes to store the result, but we don't know how much more space we need so I arbitrarily picked `* 2` as it provides plenty of headroom while still being cheaper than than `len(s)`, which requires at least `* 3` bytes and two allocs - see [`Builder.grow()`](https://github.com/golang/go/blob/release-branch.go1.13/src/strings/builder.go#L64-L70).
>
> A lower value (say `* 1.5`) could be used but one challenge is that escape sequences require either 2 or 4 bytes.  `* 2` is pretty aggressive when escaping with 2 bytes, but doesn't leave a lot of headroom when escaping with 4 bytes.

https://github.com/miekg/dns/pull/1009#discussion_r326872444

Reply:
thank you